vllm/attention/backends/rocm_flash_attn.py

```diff
@@ -441,3 +456,32 @@ def _naive_masked_attention(
     attn_weights = torch.softmax(attn_weights, dim=-1).to(value.dtype)
     out = torch.einsum("hqk,khd->qhd", attn_weights, value)
     return out
+
+
+def _sdpa_attention(query, key, value, prefill_meta, attn_metadata, num_tokens,
+                    num_heads, head_size, scale):
+    att_masks = [None] * len(prefill_meta.seq_lens)
+    attn_metadata.attn_bias = att_masks
+
+    start = 0
+    output = torch.empty((num_tokens, num_heads, head_size),
+                         dtype=query.dtype,
+                         device=query.device)
+
+    for seq_len, mask in zip(prefill_meta.seq_lens, attn_metadata.attn_bias):
+        end = start + seq_len
+        with torch.backends.cuda.sdp_kernel(enable_math=True,
+                                            enable_flash=False,
+                                            enable_mem_efficient=False):
```

Comment:
QQ: Can this completely replace the naive attention? In which case the naive attention could be better than torch SDPA?

Also, I'm wondering how it's different from the Triton and CK FlashAttention kernels.

Reply:
Yes, I believe we can replace it completely. Will update.

Reply:
The Pytorch sdpa "math" backend does not use CK Flash Attention or Triton FA.
The latest Pytorch supports Triton FA backend, and we will check later to see whether/how to integrate to vllm.